File: web/web-app/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of, throwError } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { ErrorHandlerService } from './error-handler.service';

export interface User {
    firebase_uid: string; // Firebase user identifier - single source of truth
    email: string;
    username: string;
    first_name?: string;
    last_name?: string;
    display_name?: string;
    mobile?: string;
    rating?: number;
    profile_picture?: string;
    email_verified?: boolean;
}

export interface UserRole {
    role_id?: string;
    user_id: string;
    role_name: string;
}

export interface UserClub {
    membership_id?: string;
    user: User;
    club: Club;
    role: UserRole;
}

export interface Club {
    club_id?: string;
    name: string;
    website?: string;
}

export interface CachedUserData {
    firebase_uid: string;
    email: string;
    username: string;
    display_name?: string;
    roles: string[];
    club_memberships: ClubMembership[];
    cached_at: string;
    expires_at: string;
}

export interface ClubMembership {
    club_id: string;
    club_name: string;
    role: string;
    is_admin: boolean;
}

@Injectable({
    providedIn: 'root'
})
export class UserService {
    private apiUrl = `${environment.quarkusApiUrl}/users`;

    constructor(
        private http: HttpClient,
        private errorHandlerService: ErrorHandlerService
    ) {}

    /**
     * Get all active users
     */
    getAllUsers(): Observable<User[]> {
        return this.http.get<User[]>(this.apiUrl).pipe(
            catchError(this.handleError<User[]>('getAllUsers', []))
        );
    }

    /**
     * Get users with pagination
     */
    getUsersWithPagination(page: number, size: number): Observable<User[]> {
        return this.http.get<User[]>(`${this.apiUrl}/page/${page}/size/${size}`).pipe(
            catchError(this.handleError<User[]>('getUsersWithPagination', []))
        );
    }



    /**
     * Get a user by Firebase UID
     */
    getUserByFirebaseUid(firebaseUid: string): Observable<User | null> {
        return this.http.get<User>(`${this.apiUrl}/firebase/${firebaseUid}`).pipe(
            catchError(this.handleError<User | null>('getUserByFirebaseUid', null))
        );
    }

    /**
     * Get a user by email
     */
    getUserByEmail(email: string): Observable<User | null> {
        return this.http.get<User>(`${this.apiUrl}/email/${email}`).pipe(
            catchError(this.handleError<User | null>('getUserByEmail', null))
        );
    }

    /**
     * Search users by name, email, or username
     */
    searchUsers(searchTerm: string): Observable<User[]> {
        return this.http.get<User[]>(`${this.apiUrl}/search?q=${encodeURIComponent(searchTerm)}`).pipe(
            catchError(this.handleError<User[]>('searchUsers', []))
        );
    }

    /**
     * Create a new user
     */
    createUser(user: User): Observable<User | null> {
        return this.http.post<User>(this.apiUrl, user).pipe(
            catchError(this.handleError<User | null>('createUser', null))
        );
    }



    /**
     * Update user profile by Firebase UID
     */
    updateUserProfile(firebaseUid: string, profileData: any): Observable<User | null> {
        return this.http.put<User>(`${this.apiUrl}/firebase/${firebaseUid}/profile`, profileData).pipe(
            catchError(this.handleError<User | null>('updateUserProfile', null))
        );
    }



    /**
     * Get all roles for a user by Firebase UID
     */
    getUserRoles(firebaseUid: string): Observable<UserRole[]> {
        return this.http.get<UserRole[]>(`${this.apiUrl}/firebase/${firebaseUid}/roles`).pipe(
            catchError(this.handleError<UserRole[]>('getUserRoles', []))
        );
    }

    /**
     * Get all roles for a user by Firebase UID (keycloak UID)
     */
    getUserRolesByCognitoSub(cognitoSub: string): Observable<UserRole[]> {
        // Call backend endpoint that resolves roles directly by Cognito sub
        return this.http.get<UserRole[]>(`${this.apiUrl}/cognito/${cognitoSub}/roles`).pipe(
            catchError(this.handleError<UserRole[]>('getUserRolesByCognitoSub', []))
        );
    }

    /**
     * Add a role to a user by Firebase UID
     */
    addRoleToUser(firebaseUid: string, roleName: string): Observable<UserRole | null> {
        return this.http.post<UserRole>(`${this.apiUrl}/firebase/${firebaseUid}/roles?role=${roleName}`, {}).pipe(
            catchError(this.handleError<UserRole | null>('addRoleToUser', null))
        );
    }

    /**
     * Add a role to a user by Cognito sub
     */
    addRoleToUserByCognitoSub(cognitoSub: string, roleName: string): Observable<UserRole | null> {
        // Use backend endpoint that adds role using Cognito sub directly
        return this.http.post<UserRole>(`${this.apiUrl}/cognito/${cognitoSub}/roles?role=${encodeURIComponent(roleName)}`, {}).pipe(
            catchError(this.handleError<UserRole | null>('addRoleToUserByCognitoSub', null))
        );
    }

    /**
     * Add a role to a user by Firebase UID (keycloak UID)
     */
    addRoleToUserByFirebaseUid(firebaseUid: string, roleName: string): Observable<UserRole | null> {
        return this.getUserByFirebaseUid(firebaseUid).pipe(
            switchMap(user => {
                if (!user || !user.firebase_uid) {
                    throw new Error('User not found or firebase_uid is missing');
                }
                return this.addRoleToUser(user.firebase_uid, roleName);
            }),
            catchError(this.handleError<UserRole | null>('addRoleToUserByFirebaseUid', null))
        );
    }

    /**
     * Remove a role from a user by Firebase UID
     */
    removeRoleFromUser(firebaseUid: string, roleName: string): Observable<boolean> {
        return this.http.delete(`${this.apiUrl}/firebase/${firebaseUid}/roles/${roleName}`).pipe(
            map(() => true),
            catchError(this.handleError<boolean>('removeRoleFromUser', false))
        );
    }

    /**
     * Check if a user has a specific role
     */
    userHasRole(firebaseUid: string, roleName: string): Observable<boolean> {
        return this.getUserRoles(firebaseUid).pipe(
            map(roles => roles.some(role => role.role_name === roleName)),
            catchError(this.handleError<boolean>('userHasRole', false))
        );
    }

    /**
     * Get all clubs for a user by Firebase UID
     */
    getUserClubs(firebaseUid: string): Observable<UserClub[]> {
        return this.http.get<UserClub[]>(`${this.apiUrl}/firebase/${firebaseUid}/clubs`).pipe(
            catchError(this.handleError<UserClub[]>('getUserClubs', []))
        );
    }

    /**
     * Get all clubs for a user by Firebase UID
     */
    getUserClubsByCognitoSub(cognitoSub: string): Observable<UserClub[]> {
        return this.http.get<UserClub[]>(`${this.apiUrl}/cognito/${cognitoSub}/clubs`).pipe(
            catchError(this.handleError<UserClub[]>('getUserClubsByCognitoSub', []))
        );
    }

    /**
     * Add a user to a club by Firebase UID
     */
    addUserToClub(firebaseUid: string, clubId: string, role: string = 'member'): Observable<UserClub | null> {
        return this.http.post<UserClub>(`${this.apiUrl}/firebase/${firebaseUid}/clubs/${clubId}?role=${role}`, {}).pipe(
            catchError(this.handleError<UserClub | null>('addUserToClub', null))
        );
    }

    /**
     * Add a user to a club by Firebase UID
     */
    addUserToClubByFirebaseUid(firebaseUid: string, clubId: string, role: string = 'member'): Observable<UserClub | null> {
        return this.getUserByFirebaseUid(firebaseUid).pipe(
            switchMap((user: User | null) => {
                if (!user || !user.firebase_uid) {
                    return of(null);
                }
                return this.addUserToClub(user.firebase_uid, clubId, role);
            }),
            catchError(this.handleError<UserClub | null>('addUserToClubByFirebaseUid', null))
        );
    }

    /**
     * Remove a user from a club by Firebase UID
     */
    removeUserFromClub(firebaseUid: string, clubId: string): Observable<boolean> {
        return this.http.delete(`${this.apiUrl}/firebase/${firebaseUid}/clubs/${clubId}`).pipe(
            map(() => true),
            catchError(this.handleError<boolean>('removeUserFromClub', false))
        );
    }

    /**
     * Remove a user from a club by Firebase UID
     */
    removeUserFromClubByFirebaseUid(firebaseUid: string, clubId: string): Observable<boolean> {
        return this.removeUserFromClub(firebaseUid, clubId);
    }

    /**
     * Get or create user from Firebase profile
     */
    getOrCreateUserFromFirebase(firebaseProfile: any): Observable<User | null> {
        const firebaseUid = firebaseProfile.firebase_uid;
        
        if (!firebaseUid) {
            console.error('No firebase_uid provided in firebaseProfile:', firebaseProfile);
            return of(null);
        }
        
        return this.getUserByFirebaseUid(firebaseUid).pipe(
            switchMap(existingUser => {
                if (existingUser) {
                    return of(existingUser);
                } else {
                    // Create new user
                    const user: User = {
                        firebase_uid: firebaseUid,
                        email: firebaseProfile.email,
                        username: firebaseProfile.display_name || firebaseProfile.email?.split('@')[0],
                        first_name: firebaseProfile.first_name || firebaseProfile.display_name?.split(' ')[0],
                        last_name: firebaseProfile.last_name || firebaseProfile.display_name?.split(' ').slice(1).join(' '),
                        display_name: firebaseProfile.display_name || `${firebaseProfile.first_name || ''} ${firebaseProfile.last_name || ''}`.trim(),
                        email_verified: firebaseProfile.email_verified || false
                    };
                    return this.createUser(user);
                }
            }),
            catchError(this.handleError<User | null>('getOrCreateUserFromFirebase', null))
        );
    }

    /**
     * Health check
     */
    healthCheck(): Observable<string> {
        return this.http.get(`${this.apiUrl}/health`, { responseType: 'text' }).pipe(
            catchError(this.handleError<string>('healthCheck', 'Service unavailable'))
        );
    }

    /**
     * Get cached user authentication data (roles and club memberships)
     */
    getCachedUserAuthData(firebaseUid: string): Observable<CachedUserData> {
        return this.http.get<CachedUserData>(`${this.apiUrl}/firebase/${firebaseUid}/auth-data`).pipe(
            catchError(this.handleError<CachedUserData>('getCachedUserAuthData'))
        );
    }

    /**
     * Invalidate user cache
     */
    invalidateUserCache(firebaseUid: string): Observable<string> {
        return this.http.delete(`${this.apiUrl}/firebase/${firebaseUid}/cache`, { responseType: 'text' }).pipe(
            catchError(this.handleError<string>('invalidateUserCache', 'Cache invalidation failed'))
        );
    }

    /**
     * Get user's admin clubs
     */
    getUserAdminClubs(firebaseUid: string): Observable<ClubMembership[]> {
        return this.http.get<ClubMembership[]>(`${this.apiUrl}/firebase/${firebaseUid}/admin-clubs`).pipe(
            catchError(this.handleError<ClubMembership[]>('getUserAdminClubs', []))
        );
    }

    /**
     * Handle HTTP errors
     */
    private handleError<T>(operation = 'operation', result?: T) {
        return (error: any): Observable<T> => {
            // Notify user of the error
            this.errorHandlerService.handleApiError(error, operation);
            
            // For read-only operations, return default value
            // For write operations, rethrow the error
            const isReadOperation = operation.includes('get') || operation.includes('search') || operation.includes('health');
            
            if (isReadOperation && result !== undefined) {
                return of(result as T);
            } else {
                return throwError(() => error);
            }
        };
    }
} 

File: services/src/main/java/za/cf/cp/user/service/UserService.java
package za.cf.cp.user.service;

import io.quarkus.panache.common.Sort;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import za.cf.cp.user.User;
import za.cf.cp.user.UserRole;
import za.cf.cp.user.Role;
import za.cf.cp.user.UserClub;
import za.cf.cp.club.Club;
import za.cf.cp.user.dto.CachedUserData;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service class for managing users in PostgreSQL.
 * Uses firebase_uid as the primary identifier.
 */
@ApplicationScoped
public class UserService {
    
    @Inject
    UserAuthCacheService userAuthCacheService;
    
    /**
     * Find a user by their Firebase UID
     */
    public Optional<User> findByFirebaseUid(String firebaseUid) {
        return User.find("firebaseUid", firebaseUid).firstResultOptional();
    }
    
    /**
     * Find a user by their email
     */
    public Optional<User> findByEmail(String email) {
        return User.find("email", email).firstResultOptional();
    }
    
    /**
     * Find a user by their username
     */
    public Optional<User> findByUsername(String username) {
        return User.find("username", username).firstResultOptional();
    }
    
    /**
     * Find a user by their Firebase UID (alias for findByFirebaseUid)
     */
    public Optional<User> findById(String firebaseUid) {
        return findByFirebaseUid(firebaseUid);
    }
    
    /**
     * Get all users
     */
    public List<User> findAll() {
        return User.findAll().list();
    }
    
    /**
     * Get all users with pagination
     */
    public List<User> findAllWithPagination(int page, int size) {
        return User.findAll(Sort.by("firebaseUid").descending())
                .page(page, size)
                .list();
    }
    
    /**
     * Create a new user
     */
    @Transactional
    public User createUser(User user) {
        // Validate required fields
        if (user.firebaseUid == null || user.firebaseUid.trim().isEmpty()) {
            throw new RuntimeException("Firebase UID is required");
        }
        if (user.email == null || user.email.trim().isEmpty()) {
            throw new RuntimeException("Email is required");
        }
        if (user.username == null || user.username.trim().isEmpty()) {
            throw new RuntimeException("Username is required");
        }
        
        // Apply sensible defaults
        if (user.displayName == null || user.displayName.trim().isEmpty()) {
            user.displayName = user.username;
        }
        if ((user.firstName == null || user.firstName.isBlank()) && (user.lastName == null || user.lastName.isBlank())) {
            // Derive first/last from displayName if possible (split on first space)
            String dn = user.displayName != null ? user.displayName.trim() : "";
            int spaceIdx = dn.indexOf(' ');
            if (spaceIdx > 0) {
                user.firstName = dn.substring(0, spaceIdx).trim();
                user.lastName = dn.substring(spaceIdx + 1).trim();
            } else {
                user.firstName = dn; // put all in first name
            }
        }
        if (user.rating == null) {
            user.rating = 0;
        }
        if (user.emailVerified == null) {
            user.emailVerified = false;
        }
        if (user.interests == null) {
            user.interests = new String[0];
        }
        if (user.profileCompleted == null) {
            user.profileCompleted = false;
        }
        
        // Ensure username is unique by appending a suffix if needed
        String baseUsername = user.username;
        String uniqueUsername = baseUsername;
        int suffix = 1;
        
        while (findByUsername(uniqueUsername).isPresent()) {
            uniqueUsername = baseUsername + suffix;
            suffix++;
        }
        user.username = uniqueUsername;
        
        user.persist();
        
        // Assign default role 'player' if no roles exist yet
        assignDefaultPlayerRole(user.firebaseUid);
        
        return user;
    }
    
    private void assignDefaultPlayerRole(String firebaseUid) {
        try {
            // Check if user already has any role
            List<UserRole> roles = UserRole.find("user.firebaseUid", firebaseUid).list();
            
            if (roles == null || roles.isEmpty()) {
                Optional<User> userOpt = findByFirebaseUid(firebaseUid);
                if (userOpt.isPresent()) {
                    // Find the 'player' role from the role table
                    Role playerRole = Role.find("roleName", "player").firstResult();
                    if (playerRole == null) {
                        System.err.println("❌ 'player' role not found in role table");
                        return;
                    }
                    
                    UserRole userRole = new UserRole(userOpt.get(), playerRole);
                    userRole.persist();
                }
            }
        } catch (Exception e) {
            System.err.println("❌ Error creating default role: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Update an existing user
     */
    @Transactional
    public User updateUser(User user) {
        User existingUser = User.findById(user.firebaseUid);
        if (existingUser == null) {
            throw new RuntimeException("User not found with Firebase UID: " + user.firebaseUid);
        }
        
        // Update fields
        existingUser.firstName = user.firstName;
        existingUser.lastName = user.lastName;
        existingUser.displayName = user.displayName;
        existingUser.mobile = user.mobile;
        existingUser.rating = user.rating;
        existingUser.profilePicture = user.profilePicture;
        existingUser.emailVerified = user.emailVerified;
        if (user.interests != null) existingUser.interests = user.interests;
        if (user.profileCompleted != null) existingUser.profileCompleted = user.profileCompleted;
        existingUser.playtomicRating = user.playtomicRating;
        
        return existingUser;
    }
    
    /**
     * Delete a user
     */
    @Transactional
    public boolean deleteUser(String firebaseUid) {
        User user = User.findById(firebaseUid);
        if (user != null) {
            user.delete();
            return true;
        }
        return false;
    }
    
    /**
     * Get user roles by Firebase UID
     */
    public List<UserRole> getUserRoles(String firebaseUid) {
        return UserRole.find("user.firebaseUid", firebaseUid).list();
    }
    
    /**
     * Get user roles by Firebase UID (alias for getUserRoles)
     */
    public List<UserRole> getUserRolesByFirebaseUid(String firebaseUid) {
        return getUserRoles(firebaseUid);
    }
    
    /**
     * Add role to user
     */
    @Transactional
    public UserRole addRoleToUser(String firebaseUid, String roleName) {
        Optional<User> userOpt = findByFirebaseUid(firebaseUid);
        if (userOpt.isEmpty()) {
            throw new RuntimeException("User not found with Firebase UID: " + firebaseUid);
        }
        
        // Find the role from the role table
        Role role = Role.find("roleName", roleName).firstResult();
        if (role == null) {
            throw new RuntimeException("Role not found: " + roleName);
        }
        
        User user = userOpt.get();
        UserRole userRole = new UserRole(user, role);
        userRole.persist();
        return userRole;
    }
    
    /**
     * Add role to user by Firebase UID (alias for addRoleToUser)
     */
    @Transactional
    public UserRole addRoleToUserByFirebaseUid(String firebaseUid, String roleName) {
        return addRoleToUser(firebaseUid, roleName);
    }
    
    /**
     * Remove role from user
     */
    @Transactional
    public boolean removeRoleFromUser(String firebaseUid, String roleName) {
        List<UserRole> userRoles = UserRole.find("user.firebaseUid = ?1 and roleName = ?2", firebaseUid, roleName).list();
        if (!userRoles.isEmpty()) {
            userRoles.get(0).delete();
            return true;
        }
        return false;
    }
    
    /**
     * Get user club memberships by Firebase UID
     */
    public List<UserClub> getUserClubs(String firebaseUid) {
        return UserClub.find("user.firebaseUid", firebaseUid).list();
    }
    
    /**
     * Add user to club
     */
    @Transactional
    public UserClub addUserToClub(String firebaseUid, UUID clubId, String roleName) {
        Optional<User> userOpt = findByFirebaseUid(firebaseUid);
        if (userOpt.isEmpty()) {
            throw new RuntimeException("User not found with Firebase UID: " + firebaseUid);
        }
        
        Club club = Club.findById(clubId);
        if (club == null) {
            throw new RuntimeException("Club not found with ID: " + clubId);
        }
        
        // Find the role by name
        Role role = Role.find("roleName", roleName).firstResult();
        if (role == null) {
            throw new RuntimeException("Role not found: " + roleName);
        }
        
        User user = userOpt.get();
        UserClub userClub = new UserClub(user, club, role);
        userClub.persist();
        return userClub;
    }
    
    /**
     * Remove user from club
     */
    @Transactional
    public boolean removeUserFromClub(String firebaseUid, UUID clubId) {
        List<UserClub> userClubs = UserClub.find("user.firebaseUid = ?1 and club.clubId = ?2", firebaseUid, clubId).list();
        if (!userClubs.isEmpty()) {
            userClubs.get(0).delete();
            return true;
        }
        return false;
    }
    
    /**
     * Get club admins by club ID
     */
    public List<UserClub> getClubAdmins(UUID clubId) {
        return UserClub.find("club.clubId = ?1 and role.roleName in ('admin', 'owner')", clubId).list();
    }
    
    
    /**
     * Get user's club memberships with roles
     */
    public List<UserClub> getUserClubMemberships(String firebaseUid) {
        return UserClub.find("user.firebaseUid = ?1", firebaseUid).list();
    }
    
    /**
     * Get cached user authentication data (roles and club memberships)
     */
    public CachedUserData getCachedUserData(String firebaseUid) {
        return userAuthCacheService.getCachedUserData(firebaseUid);
    }
    
    /**
     * Check if user has a specific role (using cache)
     */
    public boolean hasRole(String firebaseUid, String roleName) {
        return userAuthCacheService.hasRole(firebaseUid, roleName);
    }
    
    /**
     * Check if user is admin of a specific club (using cache)
     */
    public boolean isUserClubAdmin(String firebaseUid, UUID clubId) {
        return userAuthCacheService.isAdminOfClub(firebaseUid, clubId);
    }
    
    /**
     * Get all clubs where user is admin (using cache)
     */
    public List<CachedUserData.ClubMembership> getAdminClubs(String firebaseUid) {
        return userAuthCacheService.getAdminClubs(firebaseUid);
    }
    
    /**
     * Get all club memberships for user (using cache)
     */
    public List<CachedUserData.ClubMembership> getCachedClubMemberships(String firebaseUid) {
        return userAuthCacheService.getClubMemberships(firebaseUid);
    }
    
    /**
     * Invalidate user cache (call when user data changes)
     */
    public void invalidateUserCache(String firebaseUid) {
        userAuthCacheService.invalidateUserCache(firebaseUid);
    }
    
    /**
     * Check if user exists by Firebase UID
     */
    public boolean existsByFirebaseUid(String firebaseUid) {
        return User.count("firebaseUid", firebaseUid) > 0;
    }
    
    /**
     * Check if user exists by email
     */
    public boolean existsByEmail(String email) {
        return User.count("email", email) > 0;
    }
    
    /**
     * Check if user exists by username
     */
    public boolean existsByUsername(String username) {
        return User.count("username", username) > 0;
    }
    
    /**
     * Search users by name, email, or username
     */
    public List<User> searchUsers(String searchTerm) {
        if (searchTerm == null || searchTerm.trim().isEmpty()) {
            return findAll();
        }
        
        String trimmedSearchTerm = "%" + searchTerm.trim().toLowerCase() + "%";
        
        return User.find("""
            LOWER(firstName) LIKE ?1 OR 
            LOWER(lastName) LIKE ?1 OR 
            LOWER(displayName) LIKE ?1 OR 
            LOWER(email) LIKE ?1 OR 
            LOWER(username) LIKE ?1
            """, trimmedSearchTerm).list();
    }
} 